File: Samples/CSharp/KeyRecorder/Form1.cs
using System;

using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using OpenNETCF.Windows.Forms;
using OpenNETCF.Win32;

namespace KeyRecorder
{
  public partial class Form1 : Form
  {
    private KeyboardHook m_keyHook;
    private int m_hideshift = 0;
    private bool m_hidden = false;

    
    public Form1()
    {
      m_keyHook = new KeyboardHook();

      InitializeComponent();

      clear.Click += clear_Click;
      hideShow.Click += hideShow_Click;

      m_keyHook.KeyDetected += OnKeyDetected;
      m_keyHook.Enabled = true;

      inputPanel.EnabledChanged += inputPanel_EnabledChanged;
    }

    void hideShow_Click(object sender, EventArgs e)
    {
      m_hidden = !m_hidden;

      if (m_hidden)
      {
        hideShow.BackColor = SystemColors.ActiveBorder;
        hideShow.Top = 0;
      }
      else
      {
        hideShow.BackColor = SystemColors.Control;
        hideShow.Top = 4;
      }

      m_hideshift = m_hideshift == 0 ? 27 : 0;
      Move();
    }

    void inputPanel_EnabledChanged(object sender, EventArgs e)
    {
      Move();
    }

    private void Move()
    {
      if (inputPanel.Enabled)
      {
        this.Top = inputPanel.Bounds.Top - 30 + m_hideshift;
      }
      else
      {
        this.Top = Screen.PrimaryScreen.WorkingArea.Height - 30 + m_hideshift;
      }
    }

    protected override void OnLoad(EventArgs e)
    {
      this.Width = Screen.PrimaryScreen.WorkingArea.Width;
      this.Height = 29;
      Move();

      Win32Window keysWindow = new Win32Window(keys.Handle);
      keysWindow.ExtendedStyle |= WS_EX.NOACTIVATE;

      base.OnLoad(e);
    }

    protected override void OnPaintBackground(PaintEventArgs e)
    {
      e.Graphics.FillRectangle(new SolidBrush(SystemColors.ControlDark), 0, 0, this.Width, this.Height);
    }

    void clear_Click(object sender, EventArgs e)
    {
      keys.Text = string.Empty;
    }

    void OnKeyDetected(OpenNETCF.Win32.WM keyMessage, KeyData keyData)
    {
      if (this.InvokeRequired)
      {
        this.Invoke(new KeyHookEventHandler(OnKeyDetected), new object[] { keyMessage, keyData });
        return;
      }
      char c = (char)keyData.KeyCode;
      if (keyMessage == WM.KEYUP)
      {
        if (char.IsLetterOrDigit(c))
        {
          keys.Text += c;
        }
        else
        {
          switch ((Keys)c)
          {
            case Keys.Up:
              keys.Text += "{up}";
              break;
            case Keys.Down:
              keys.Text += "{dn}";
              break;
            case Keys.Left:
              keys.Text += "{lt}";
              break;
            case Keys.Right:
              keys.Text += "{rt}";
              break;
          }
        }
      }
      keys.SelectionStart = keys.Text.Length;
      keys.ScrollToCaret();
    }
  }
}

File: Samples/VB/KeyRecorder/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports OpenNETCF.Windows.Forms
Imports OpenNETCF.Win32

Namespace KeyRecorder

    Public Class Form1
        Dim m_keyHook As KeyboardHook = New KeyboardHook()
        Dim m_hideshift As Integer = 0
        Dim m_hidden As Boolean = False
    
        Private Sub btnHideShow_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnHideShow.Click
            m_hidden = Not m_hidden
            
            If m_hidden Then                
                btnHideShow.BackColor = SystemColors.ActiveBorder
                btnHideShow.Top = 0
            Else
                btnHideShow.BackColor = SystemColors.Control
                btnHideShow.Top = 4
            End If

            m_hideshift = IIf(m_hideshift = 0, 27, 0)
            Move()
        End Sub

        Private Sub InputPanel1_EnabledChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles InputPanel1.EnabledChanged
            Move()
        End Sub
        
        Private Sub Move()
            If InputPanel1.Enabled Then
                Me.Top = InputPanel1.Bounds.Top - 30 + m_hideshift
            Else
                Me.Top = Screen.PrimaryScreen.WorkingArea.Height - 30 + m_hideshift
            End If
        End Sub

        Protected Overloads Overrides Sub OnLoad(ByVal e As EventArgs)
            m_keyHook.Enabled = True
            AddHandler m_keyHook.KeyDetected, AddressOf OnKeyDetected
            Me.Width = Screen.PrimaryScreen.WorkingArea.Width
            Me.Height = 29
            Move()

            Dim keysWindow As New Win32Window(txtKeys.Handle)
            keysWindow.ExtendedStyle = keysWindow.ExtendedStyle Or WS_EX.NOACTIVATE

            MyBase.OnLoad(e)
        End Sub

        Protected Overloads Overrides Sub OnPaintBackground(ByVal e As PaintEventArgs)
            e.Graphics.FillRectangle(New SolidBrush(SystemColors.ControlDark), 0, 0, Me.Width, Me.Height)
        End Sub

        Private Sub btnClear_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClear.Click
            txtKeys.Text = String.Empty
        End Sub

        Private Sub OnKeyDetected(ByVal keyMessage As OpenNETCF.Win32.WM, ByVal keyData As KeyData)
            If Me.InvokeRequired Then
                Me.Invoke(New KeyHookEventHandler(AddressOf OnKeyDetected), New Object() {keyMessage, keyData})
                Return
            End If
            Dim c As Char = CChar(CStr(keyData.KeyCode))
            If keyMessage = WM.KEYUP Then
                If Char.IsLetterOrDigit(c) Then
                    txtKeys.Text += c
                Else
                    Select Case CType(AscW(c), Keys)
                        Case Keys.Up
                            txtKeys.Text += "{up}"
                            Exit Select
                        Case Keys.Down
                            txtKeys.Text += "{dn}"
                            Exit Select
                        Case Keys.Left
                            txtKeys.Text += "{lt}"
                            Exit Select
                        Case Keys.Right
                            txtKeys.Text += "{rt}"
                            Exit Select
                    End Select
                End If
            End If
            txtKeys.SelectionStart = txtKeys.Text.Length
            txtKeys.ScrollToCaret()
        End Sub        
    End Class

End Namespace